kvserver/rangefeed: move rangefeed level cleanup to stream muxer

wenyihu6 opened the pull request:
Previously, each rangefeed had a dedicated goroutine to manage the output loop,
disconnects, and cleanups. In an upcoming commit, we aim to remove these
goroutines. This patch shifts the cleanup responsibilities to the stream muxer
to make future commits cleaner.

Now, each rangefeed can register a clean up function with the stream muxer,
which promises to invoke these callbacks during stream disconnects. Note that we
intend to eliminate the O(ranges) goroutines only for scheduled processors. The
the legacy processor remains unchanged by not registering any clean up functions.

Part of:
Epic: none
Release note: none

wenyihu6 committed:
kvserver/rangefeed: remove lockedRangefeedStream

Previously, we created separate locked rangefeed streams for each individual
rangefeed stream to ensure Send can be called concurrently as the underlying
grpc stream is not thread safe. However, since the introduction of the mux
rangefeed support, we already have a dedicated lock for the underlying mux
stream, making the Send method on each rangefeed stream thread safe already.
This patch removes the redundant locks from each individual rangefeed stream.

Epic: none
Release note: none

wenyihu6 committed:
kvserver/rangefeed: remove non-mux rangefeed metrics

Previously, we removed non-mux rangefeed code in
https://github.com/cockroachdb/cockroach/pull/125610. However, that patch forgot
to remove non-mux rangefeed metrics. This patch removes these metrics as they
are no longer needed.

Epic: none
Release note: none

wenyihu6 committed:
kvserver/rangefeed: introduce stream muxer

This patch introduces a new stream muxer struct to manage server rangefeed
streams. Currently, it delegates the existing rangefeed completion watcher to
stream muxer without changing any existing behavior. The main purpose is to make
future commits cleaner.

Part of:
Epic: none
Release note: none

wenyihu6 committed:
kvserver/rangefeed: move active streams to stream muxer

This patch delegates the existing active stream management to stream muxer
without changing any existing behavior. The main purpose is to make future
commits cleaner.

Part of:
Epic: none
Release note: none

wenyihu6 committed:
kvserver/rangefeed: move metrics updates to stream muxer

This patch delegates the existing rangefeed metrics management to stream muxer
without changing any existing behavior. The main purpose is to make future
commits cleaner.

Part of:
Epic: none
Release note: none

wenyihu6 committed:
kvserver/rangefeed: remove unused cancel

This patch removes the unused field cancel from setRangeIDEventSink.

Part of:
Epic: none
Release note: none

wenyihu6 committed:
kvserver/rangefeed: wrap muxer inside setRangeIDEventSink

Previously, each rangefeed stream setRangeIDEventSink wrapped the underlying
muxstream to send to grpc stream. This patch changes the stream to wrap the
stream muxer struct instead which itself wraps the underlying muxstream.

Note that this patch doesn’t change any existing behavior but only to make
future commits cleaner when we change muxer to send data to a buffered stream
instead of directly to grpc.

Part of:
Epic: none
Release note: none

wenyihu6 committed:
kvserver/rangefeed: remove future package

Previously, to reduce O(ranges) goroutines and avoid blocking on rangefeed
completion, we introduced the future package to manage rangefeed completion by
invoking a callback when a future error becomes ready. This patch replicas it
with a new approach - introduce a `DisconnectRangefeedWithError` method on
stream muxer. Each rangefeed stream can now call this function to signal
rangefeed completion, allowing the muxer to handle shutdown logic.

Part of:
Epic: none
Release note: none

wenyihu6 committed:
kvserver/rangefeed: add rangeID as part of activeStreams map

This patch adds rangeID into the activeStreams map without changing any existing
behavior. The main purpose is to make future commits cleaner.

Part of:
Epic: none
Release note: none

wenyihu6 committed:
kvserver/rangefeed: move rangefeed level cleanup to stream muxer

Previously, each rangefeed had a dedicated goroutine to manage the output loop,
disconnects, and cleanups. In an upcoming commit, we aim to remove these
goroutines. This patch shifts the cleanup responsibilities to the stream muxer
to make future commits cleaner.

Now, each rangefeed can register a clean up function with the stream muxer,
which promises to invoke these callbacks during stream disconnects. Note that we
intend to eliminate the O(ranges) goroutines only for scheduled processors. The
the legacy processor remains unchanged by not registering any clean up functions.

Part of:
Epic: none
Release note: none

wenyihu6 committed:
to be squashed - test changes separated out

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126492)
<!-- Reviewable:end -->

wenyihu6 commented:
*[`pkg/kv/kvserver/rangefeed/stream_muxer.go` line 180 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/126492#-O0iY89T1HIbHapEoV_B:-O0iY89T1HIbHapEoV_C:baw5a2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd40802393fad132e2ea639d56be8415ed2220e/pkg/kv/kvserver/rangefeed/stream_muxer.go#L180)):*
> ```Go
> // Only muxer.Run can call this function, so it is fine to be blocking and
> // invoking clean up function directly.
> func (sm *StreamMuxer) DisconnectAllWithErr(err *kvpb.Error) {
> ```

When the stream muxer is stopped, I could think of two approaches for shutting down corresponding streams. One approach is to return an error on every Send method, allowing rangefeed disconnect itself. The other one is to iterate through active streams and let muxer shut down all streams. The first approach is better for letting rangefeeds manage itself, but it could delay muxer's termination since it has to wait for the next Send() to invoke the clean up function for every registration. In the future commit when I introduce buffered streams (plan: https://github.com/cockroachdb/cockroach/blob/be2cc8f904cc5201e76e6a503d2e514cc26e1d73/pkg/server/node.go#L1904), we will maintain the same strategy. Buffered stream send never returns an error (even when overflowed). Instead, it will send a shutdown signal to the muxer which will handle disconnection. 


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/kv/kvserver/rangefeed/stream_muxer.go` line 180 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/126492#-O0iY89T1HIbHapEoV_B:-O0iYdASEehNShrt1GrG:b-gfp40y) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd40802393fad132e2ea639d56be8415ed2220e/pkg/kv/kvserver/rangefeed/stream_muxer.go#L180)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

When the stream muxer is stopped, I could think of two approaches for shutting down corresponding streams. One approach is to return an error on every Send method, allowing rangefeed disconnect itself. The other one is to iterate through active streams and let muxer shut down all streams. The first approach is better for letting rangefeeds manage itself, but it could delay muxer's termination since it has to wait for the next Send() to invoke the clean up function for every registration. In the future commit when I introduce buffered streams (plan: https://github.com/cockroachdb/cockroach/blob/be2cc8f904cc5201e76e6a503d2e514cc26e1d73/pkg/server/node.go#L1904), we will maintain the same strategy. Buffered stream send never returns an error (even when overflowed). Instead, it will send a shutdown signal to the muxer which will handle disconnection.
</blockquote></details>

Lmk your thoughts. 


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126492)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/kv/kvserver/rangefeed/stream_muxer.go` line 169 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/126492#-O0i_XROCmpnlqTFrLfM:-O0i_XROCmpnlqTFrLfN:b-o0bju9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd40802393fad132e2ea639d56be8415ed2220e/pkg/kv/kvserver/rangefeed/stream_muxer.go#L169)):*
> ```Go
> 	// rely on the map rangefeedCleanUps to dedupe during Run.
> 	if _, ok := sm.rangefeedCleanUps.Load(streamID); ok {
> 		sm.appendCleanUp(streamID)
> ```

Initially, I considered using a single channel and calling cleanup functions when sending a mux error to the client. However, after some thoughts, I think it's better to keep active streams and rangefeed cleanups separate. This is because active streams and rangefeed cleanups don’t necessarily share the same set of IDs. An active stream might have an ID before registration is properly set up, resulting in the rangefeed cleanup not having the ID. Similarly, a client might request to close a stream before p.Register happens. And registration disconnects might have not happened yet before r.RegisterRangefeedCleanUp. Although I’m unsure how often this happens in practice, keeping the management separate seems more accurate.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
@nvanbenschoten I tagged you as reviewer, but feel free to defer some if it’s too many.

wenyihu6 committed:
kvserver: wrap kvpb.RangeFeedEventSink in Stream

Previously, we declared the same interface signature twice: once in
kvpb.RangeFeedEventSink and again in rangefeed.Stream. This patch embeds
kvpb.RangeFeedEventSink inside rangefeed.Stream, making rangefeed.Stream a
superset of kvpb.RangeFeedEventSink. This approach makes sense, as each
rangefeed server stream should be a rangefeed event sink, capable of making
thread-safe rangefeed event sends.

Epic: none
Release note: none

wenyihu6 commented:
Closing for now: I ended up going with a different route - I will introduce the registration interface first, and limit the remianing changes including changes in this pr to nonbuffered registrtation struct only.
